### DISCUSSION TITLE
Nested pattern matching with AsyncData<Result>

### DISCUSSION THREAD

**Discussion by drazik** [bloodyowl/boxed#88] [Q&A] [answered]
Hello,

I am trying to implement a React component that holds an `AsyncData<Result>` in a state, and use pattern matching (using ts-pattern) to render the markup based on this state.

Here is an example: [TS playground example](https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAJQKYEMDGMA0cDecCuAzkgMowoxLZFICiAZvUhnAL5z1QQhwBEUqDLwBQoSLFxwAgoQCeAOzQARCimzJC+ADZY4AMXwx8Athy49eAAUIB3FPIC0wCAHoARhAAeSACYix0PB4IBRoABbYAAqmnNx8MIQOYBSUUPIiwvT4ijDO8nAAKkiEMAAUAJS4wnA1cGgQ8iVwANo+qtjEMCrkALpwALwExGQUSKUyCsqqAHQAchAwMgDWvgA8Gtowq-ay1PJL8hA28gB8JxXlwtW1NAxMGKUVAydVtW9wnd0o43KKX9MAGQgKB8wHkAHMLld3rUDEYBNNYiBIuZgMRSkwYOFSgADFwoMDAFxtcg47DBJAwMIQHwALjgAHJICUGWxypcYTDpiEwAB5JYFCAbHSPASESCNJCVfona6czliiXEaYQJZwAD8iGKm2m-NKioaxEq9OFMGmtCgXCg+uKSqlmDl8rgHKdNURWgoAFkCXqbeLDVLnvpDMYkIjzCjuGixgbJdMAFaEBoXF2u7k+paPEkoaUvbOIiAQVNOlXyDQQLQANzGfs2uderven1UP0m-yUDRj2p07Md8tYxbY2GaPUujoE8PyISxYVK2cHtWmNmAVNbfxmkTmC2WvmwTxlcFWYBO83gKEIKx8qxcx4XbuXq4m6-I003QJBYPBe-rR5O79BELXrefaLg+s5PlML6bh28hjJuxBaMwZTsnuYp1s8IEwtO2JoT2mElmBpSmq+uqZvBSCIQ8KFwHOqj1qU+GNoex5fHAaK4NmrBAbKTFvL2vGgSus7EZuFpWqU5GUch5R7kglrQPRjGNr+YnQHANhhMAiEcJS4SfnUWjAEg8jwMQaATtxSkwvxjbTEgnhhCgRC5NWFQOvKd5wHZDlOSUwCuWOrBAA)

My problem is on the line 28. I am trying to extract the `Result.Ok` value using `P.select()`, but the compiler says that this is the `Result.Ok` itself, instead of the value it holds. I am a bit surprised. Is it really the intended behaviour? Since the extraction on line 30 returns the `Result.Error` value, it doesn't seem so, but maybe I didn't understand something.

Also, I don't know if it's really a problem with boxed, or if it's more related to ts-pattern.

**Comment by zoontek** [comment, marked as answer]
@drazik This seems to be a `ts-pattern` issue when the value is of type `any`. See [this demo](https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAbzgQQM4E8B2BjAIgQxnwBo4AxAVxgqgFNSAlW1CgG3gF84AzKCEOACIAAqgDu+TAFpgEAPQAjCAA9aAE0EBuAFChIsRHAqpaAUW7da2GKWO0AykRi04XXvyF181rbvDR4JBBCbAALUgAFVx4+AUEYVCkwQmcoTF9tbgocGFlMOAAVZhgAJgAKAEpEbTg4OTk4MMkAcxcYUJcAeQBrRvwTOBh0MDaIOEl0QeGXMtplbFowZzVxzHQAQjh0CAo4MWBWVjgTNtDCODVCfCmRuGBURug6awqax8xUeABtS6JSExgBCIAF04ABeIwmRyEWgAHjQWDwV1hTBY7Fh2W6mAgYkwtkwWJxmAAfKSygicED8AA6AByEBgaG66kqFR0bzs5ks1jKlXBxOqtVqAKp5IwlKu1IAMhB8GpgJhmqz2ULyFQaLRqe4QBFYvdaGVLDAwmUAAZyfBgYByX74U2kIK0doQNQALiEkE+glcFVeqtq1OCYB6BQgqLYMDKdFQkA+LjBAqjzFjJmpEF6AH44OH2NSekmYxA41V3TmYNTTFA+FACynaL6-f6taxCABZS356N1-lq6h0LWxXX8fW1oupgBWqCLrMbqsDHe6ZVtPdtWogEFnQrTmFREFYADcDdGIz2EG9-ccnaKKUiiNTcEWj8wI76dBeOGy3hxSF9gZ+3nQfb5MExqhEuVybtS+ztGKiJUtSER0gyTLqKQfIJnAsJgMS9LwP0zJqLCcjYZB0FgTe8GITKcoKs0aFVBhWHEtR8qKkRJHnnAUHADBFGSohD6YAaiEmKwViRr6pDHuw-KcbUIEmtJMCbnOZFlGWCF5ouIm0GJPKSRcVw9kxVJ3A8CC2hw7HEipW5qRpiGVtWZQ6XpEkVKQtBVtAxnYU5PliKEBwuEaYS0Y0rDALQmDwCY2CAdZtkBnMZzGLkh6VMQnGQSl+BpcAGWfhw2iZNk1h5IUxRkC2MDOJgfJnrU9RwB0dA8D57QuGotAgGMYguPgbXGOFkgMq1Ny0G82Bjt8tr-FeVyghCdjQs48Lire+Aos+6JINw67up8UDhd+RgEtiuKksSZScXxd64ShaiZW8-61JyFjibyDECo1QoilcsESneLG0cqJWqpQfaatqQ4gCOoVgealrWra9qGCATqhC67qCJ6MDeh+tnzsG3ShmWo5xj2FOpumcBZhpnbJmO9ZwKWO3lv5NZdszDZyVx3A1e2JPU-GAqQxqA78LDI7c3G1KTtOvMXsT+bLhhq77RufPbruB5Pmi8AYb9F7-UQgObfej4Fi+n7vrbrg-n+Kp1A02LOIMZzwNsuzYJIPA1XVWw7FAcDJLVXnASEQWKusAFOjQkegeBRCkTx5EbZRSGMqgBH0b5OEMuMOfqIlnHcbxGf8dKsqsXRcDoQKTEg2xxE2WXal3TSAlWw5FbeTWrmfQ2pAAPpedWVN80xnN7EFYk8E6YWKhFUUxZe8VOtZ2WpxXcFV4JBq9-mg-6QZauN9hpn3IgGvrlZrc5coqWfAVBpFdoQA)

**Reply by drazik** [reply]
Thank you! Also for the flatten pattern matching example :+1:. I will open an issue on ts-pattern then.
